### PR TITLE
Daniel/email

### PR DESCRIPTION
**Pull request by alined908**
- Times -> Added validation, and cucumber test, css
- Add linter/run rubocop
- Send email to participants, email cuke test

Instructions to send email
---------------------------
1.  Go to rosters page.  
2.  Might need to do rake db:reset if existing participants dont have secret_id, or just remove them and re-add emails.
3.  Put something in subject/body if you want
4.  Press send email button
5.  Email should have a link in the body redirecting to correct ranking page.  For development link will be http://localhost:3000/projects/:project_id/participants/:participant_id/ranking/edit?secret_id=secret_id

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -1,18 +1,21 @@
...
 gem 'bootstrap-datepicker-rails'
+gem 'rails', '4.2.10'
 # Use sqlite3 as the database for Active Record
 # Use SCSS for stylesheets
 group :production do
   gem 'pg', '~> 0.20'   # use PostgreSQL in production (Heroku)
   gem 'rails_12factor'  # Heroku-specific production settings
 end
+
+gem 'rubocop', require: false
```
Comment (lawrencechen14) [inline Gemfile:18]: Nice job with adding the linter!

File: app/controllers/matchings_controller.rb [hunk 2]
```diff
@@ -1,18 +1,18 @@
+# frozen_string_literal: true
+
 class MatchingsController < ApplicationController
-  before_action :set_matching_and_project, only: [:show, :edit, :update, :destroy]
+  before_action :set_matching_and_project, only: %i[show edit update destroy]
 
   # GET /project/:project_id/matching
   def show
-    if !logged_in?
+    unless logged_in?
```
Comment (lawrencechen14) [inline app/controllers/matchings_controller.rb:8]: I like the refactoring to make our code more Ruby-esque.

File: app/controllers/matchings_controller.rb [hunk 3]
```diff
@@ -131,28 +131,29 @@ def api
     require 'rest-client'
 
     input = {
```
Comment (lawrencechen14) [inline app/controllers/matchings_controller.rb:132]: Thanks for all of the refactoring! Makes our code look a lot nicer.

File: app/controllers/sessions_controller.rb [hunk 4]
```diff
@@ -1,25 +1,24 @@
...
-  
+  def new; end
+
   def create
     user = User.find_by(email: params[:session][:email].downcase)
-    
-    if user && user.authenticate(params[:session][:password])
+
+    if user&.authenticate(params[:session][:password])
       session[:user_id] = user.id
```
Comment (lawrencechen14) [inline app/controllers/sessions_controller.rb:10]: Nice ruby refactoring! Makes our codebase look a lot better.

File: app/controllers/times_controller.rb [hunk 5]
```diff
@@ -25,22 +25,36 @@ def index
...
-    hour = params[:timeslot_hour].to_i
-    minute = params[:timeslot_minute].to_i
-    @duration = hour * 60 + minute
     @project.update(duration: @duration)
   end
 
-  def no_project_date_time?(date_time, is_date)
-    @project.project_times.where(date_time: DateTime.parse(date_time), is_date: is_date).blank?
+  def validate_times(date, is_date)
+    datetime = DateTime.parse(date)
```
Comment (lawrencechen14) [inline app/controllers/times_controller.rb:32]: Nice job with the validating times part! Makes adding times a lot cleaner and less buggy.

File: app/mailers/participants_mailer.rb [hunk 6]
```diff
@@ -0,0 +1,13 @@
+class ParticipantsMailer < ApplicationMailer
+  default "Message-ID"=>"#{Digest::SHA2.hexdigest(Time.now.to_i.to_s)}@gmail.com"
+  
+  def availability_email(part_id, proj_id, address, secret_id, project_name, email_subject, email_body)
```
Comment (lawrencechen14) [inline app/mailers/participants_mailer.rb:4]: Great job on finding out how to send emails from the app!

File: app/models/participant.rb [hunk 7]
```diff
@@ -1,4 +1,7 @@
+# frozen_string_literal: true
+
 class Participant < ActiveRecord::Base
   belongs_to :project
+  has_secure_token :secret_id
```
Comment (lawrencechen14) [inline app/models/participant.rb:5]: Making secret_id have a secure token is great as it auto-populates the field with a unique value for each participant and project.

**Review by lawrencechen14 — APPROVED** [review]

File: app/controllers/participants_controller.rb [hunk 8]
```diff
@@ -16,8 +18,22 @@ def index
   # GET /participants/1
   def display
     @participants = Participant.where(project_id: params[:project_id])
+    @project = Project.find(params[:project_id])
   end
 
+  def email
+    @participants = Participant.where(project_id: params[:project_id])
```
Comment (kev-chien) [inline app/controllers/participants_controller.rb]: @project.participants is more Rails-esque

File: app/controllers/sessions_controller.rb (hunk 4 above)
Comment (kev-chien) [reply app/controllers/sessions_controller.rb:10]: woah this is OP, kind of obscure but worth learning

**Review by Aakup — APPROVED** [review]
Nice Job Daniel

File: config/environments/development.rb [hunk 9]
```diff
@@ -14,7 +16,27 @@
...
+  config.action_mailer.raise_delivery_errors = true
+  config.action_mailer.default_options = { from: 'multimeetemailer@gmail.com' }
+  config.action_mailer.default_url_options = { :host => "localhost:3000" }
+  config.action_mailer.delivery_method = :smtp
+  config.action_mailer.smtp_settings = {
+    address: 'smtp.gmail.com',
+    port: 587,
+    domain: 'example.com',
+    user_name: 'multimeetemailer@gmail.com',
+    password: 'ilove169!',
```
Comment (kev-chien) [inline config/environments/development.rb:36]: let's please hide this lol

**Review by kev-chien — CHANGES_REQUESTED** [review]
Overall good work! Linter is very helpful, and the email functionality is good. The database change, and the REST-ful URL for a participant given a secret_id is also very reasonable and well-thought out.

Just need to make our secrets (pw) secure and do some minor refactoring, also need to resolve merge conflicts.

File: config/environments/production.rb [hunk 10]
```diff
@@ -63,6 +65,7 @@
   # Ignore bad email addresses and do not raise email delivery errors.
   # Set this to true and configure the email server for immediate delivery to raise delivery errors.
   # config.action_mailer.raise_delivery_errors = false
+  config.action_mailer.default_url_options = { :host => "https://multi-meet.herokuapp.com" }
```
Comment (kev-chien) [inline config/environments/production.rb:68]: Thanks for figuring this out and making sure the code can seamlessly work for both dev environments and the actual environment!

**Review by kev-chien — APPROVED** [review]

File: config/environments/development.rb (hunk 9 above)
Comment (alined908) [reply config/environments/development.rb:36]: ya, ill include in next iter 👍

**Pull request merged**